File: ios/App/App/MagnetDetectorPlugin.swift
import Capacitor
import CoreMotion
import UIKit

/// The app's single CMMotionManager. Apple asks for one instance per app (more can
/// lower the delivery rate); SlamDetectorPlugin uses it as well. Declared here so no
/// extra file has to be added to the Xcode project.
let zktSharedMotionManager = CMMotionManager()

/// Raw magnetometer streamer for magnet lift-to-start ("Kaldırınca Başlat").
///
/// Deliberately dumb: it only turns CoreMotion samples into epoch-millisecond samples
/// and sends them to JS in small time-based batches. Every decision (plateaus, near/far,
/// lift onset) lives in client/util/magnet-start/ so it can be tested against real
/// recordings.
///
/// CMMagnetometer is the RAW field (device bias included). The calibrated field from
/// CMDeviceMotion keeps re-estimating the bias and slowly "learns" a cube resting next
/// to the phone as part of the device: measured 1229 -> 818 uT in 100 s, which inverts
/// near/far. Raw drifted 0.67 uT over 145 s.
@objc(MagnetDetectorPlugin)
public class MagnetDetectorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MagnetDetectorPlugin"
    public let jsName = "MagnetDetector"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getCapabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
    ]

    private static let version = 1

    private let motionManager = zktSharedMotionManager
    // Serial: the batcher is not thread-safe, and OperationQueue runs concurrently by default.
    private let sampleQueue: OperationQueue = {
        let q = OperationQueue()
        q.maxConcurrentOperationCount = 1
        q.name = "zkt.magnet.samples"
        return q
    }()
    // All start/stop/lifecycle state changes go through this queue.
    private let control = DispatchQueue(label: "zkt.magnet.control")

    private var batcher: MagnetBatcher?
    private var wanted = false
    private var periodUs = 10000
    private var batchMs = 50
    private var lifecycleObservers: [NSObjectProtocol] = []

    override public func load() {
        let center = NotificationCenter.default
        lifecycleObservers.append(center.addObserver(
            forName: UIApplication.willResignActiveNotification, object: nil, queue: nil
        ) { [weak self] _ in
            guard let self = self else { return }
            self.control.async { if self.wanted { self.stopUpdates() } }
        })
        lifecycleObservers.append(center.addObserver(
            forName: UIApplication.didBecomeActiveNotification, object: nil, queue: nil
        ) { [weak self] _ in
            guard let self = self else { return }
            self.control.async { if self.wanted { self.startUpdates() } }
        })
    }

    deinit {
        lifecycleObservers.forEach { NotificationCenter.default.removeObserver($0) }
    }

    @objc func getCapabilities(_ call: CAPPluginCall) {
        let available = motionManager.isMagnetometerAvailable
        call.resolve([
            "version": MagnetDetectorPlugin.version,
            "available": available,
            "uncalibrated": available,
            "name": "CMMagnetometer",
            "vendor": "Apple",
        ])
    }

    @objc func start(_ call: CAPPluginCall) {
        guard motionManager.isMagnetometerAvailable else {
            call.reject("Magnetometer not available", "UNAVAILABLE")
            return
        }
        let period = max(call.getInt("periodUs") ?? 10000, 5000)
        let batch = min(max(call.getInt("batchMs") ?? 50, 20), 500)
        control.async {
            self.periodUs = period
            self.batchMs = batch
            self.wanted = true
            self.startUpdates()
            call.resolve()
        }
    }

    @objc func stop(_ call: CAPPluginCall) {
        control.async {
            self.wanted = false
            self.stopUpdates()
            call.resolve()
        }
    }

    // Runs on `control`.
    private func startUpdates() {
        stopUpdates()
        let batcher = MagnetBatcher(
            spanMs: Double(max(batchMs - 10, 10)),
            startUptime: ProcessInfo.processInfo.systemUptime
        ) { [weak self] payload in
            self?.emit(payload)
        }
        self.batcher = batcher
        motionManager.magnetometerUpdateInterval = Double(periodUs) / 1_000_000.0
        motionManager.startMagnetometerUpdates(to: sampleQueue) { data, error in
            guard let data = data, error == nil else { return }
            batcher.add(data)
        }
    }

    // Runs on `control`.
    private func stopUpdates() {
        batcher?.cancel()
        batcher = nil
        if motionManager.isMagnetometerActive {
            motionManager.stopMagnetometerUpdates()
        }
    }

    private func emit(_ payload: [String: Any]) {
        // Listener bookkeeping (addListener/removeListener) runs on the bridge's serial
        // queue and notifyListeners reads it without a lock, so notify from that queue.
        if let bridge = self.bridge as? CapacitorBridge {
            bridge.dispatchQueue.async { [weak self] in
                self?.notifyListeners("samples", data: payload)
            }
        } else {
            notifyListeners("samples", data: payload)
        }
    }
}

/// Collects samples on the (serial) sample queue and flushes them by time. One instance
/// per start(); stop() cancels it so a late CoreMotion callback cannot leak into the
/// next stream.
private final class MagnetBatcher {
    private let spanMs: Double
    private let startUptime: TimeInterval
    private let sink: ([String: Any]) -> Void
    private let lock = NSLock()
    private var cancelled = false

    private var buf: [Double] = []
    private var firstT: Double = 0
    private var offsetSeconds: Double = 0
    private var clock: String?

    init(spanMs: Double, startUptime: TimeInterval, sink: @escaping ([String: Any]) -> Void) {
        self.spanMs = spanMs
        self.startUptime = startUptime
        self.sink = sink
    }

    func cancel() {
        lock.lock()
        cancelled = true
        lock.unlock()
    }

    private var isCancelled: Bool {
        lock.lock()
        defer { lock.unlock() }
        return cancelled
    }

    func add(_ data: CMMagnetometerData) {
        if isCancelled { return }

        // CoreMotion may hand back a cached sample from before this start.
        let ts = data.timestamp
        if ts < startUptime - 0.001 { return }

        let nowUptime = ProcessInfo.processInfo.systemUptime
        if clock == nil {
            // CMLogItem timestamps count from boot on the same clock as systemUptime; if a
            // device ever disagrees, fall back to arrival time rather than guess.
            let lag = nowUptime - ts
            clock = (lag >= 0 && lag < 0.5) ? "uptime" : "arrival"
        }

        let tMs: Double
        if clock == "uptime" {
            // One wall-clock offset per batch, not per sample.
            if buf.isEmpty {
                offsetSeconds = Date().timeIntervalSince1970 - nowUptime
            }
            tMs = (ts + offsetSeconds) * 1000.0
        } else {
            tMs = Date().timeIntervalSince1970 * 1000.0
        }

        let field = data.magneticField
        if buf.isEmpty { firstT = tMs }
        buf.append(tMs)
        buf.append(field.x)
        buf.append(field.y)
        buf.append(field.z)

        if tMs - firstT >= spanMs || buf.count >= 1024 {
            flush()
        }
    }

    private func flush() {
        guard !buf.isEmpty, !isCancelled else {
            buf.removeAll(keepingCapacity: true)
            return
        }
        let payload: [String: Any] = ["v": 1, "clk": clock ?? "arrival", "d": buf]
        buf.removeAll(keepingCapacity: true)
        sink(payload)
    }
}

File: android/app/src/main/java/com/zktimer/app/MagnetDetectorPlugin.java
package com.zktimer.app;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import org.json.JSONException;

/**
 * Raw magnetometer streamer for magnet lift-to-start ("Kaldırınca Başlat").
 *
 * Deliberately dumb: it only turns sensor events into epoch-millisecond samples and
 * sends them to JS in small time-based batches. Every decision (plateaus, near/far,
 * lift onset) lives in client/util/magnet-start/ so it can be tested against real
 * recordings.
 *
 * Only TYPE_MAGNETIC_FIELD_UNCALIBRATED is used. The calibrated sensor keeps
 * re-estimating hard-iron bias and slowly "learns" a cube that rests next to the phone
 * as part of the device, which inverts near/far after a minute or two (measured on
 * iOS's calibrated stream: 1229 -> 818 uT in 100 s). No calibrated fallback: a device
 * without the uncalibrated sensor reports available=false.
 *
 * Threading: start() runs on Capacitor's plugin thread, which has a Looper. The sensor
 * is registered with a Handler on that thread, so callbacks, batching and
 * notifyListeners all run there. Lifecycle callbacks run on the main thread and only
 * touch the sensor registration plus volatile flags.
 */
@CapacitorPlugin(name = "MagnetDetector")
public class MagnetDetectorPlugin extends Plugin implements SensorEventListener {

    private static final int VERSION = 1;
    private static final int DEFAULT_PERIOD_US = 10000;
    private static final int DEFAULT_BATCH_MS = 50;
    private static final int MAX_BATCH_SAMPLES = 256;

    private static final int CLOCK_UNKNOWN = 0;
    private static final int CLOCK_BOOT = 1;
    private static final int CLOCK_UPTIME = 2;
    private static final int CLOCK_ARRIVAL = 3;
    // A sensor timestamp this close behind "now" on a clock identifies that clock.
    private static final long CLOCK_MATCH_NS = 500_000_000L;

    private SensorManager sensorManager;
    private Sensor sensor;
    private Handler handler;

    // Set by start()/stop() on the plugin thread, read by lifecycle callbacks on the main thread.
    private volatile boolean active = false;
    // Asks the sensor thread to drop its batch and re-detect the clock on the next event.
    private volatile boolean resetPending = true;
    private volatile int periodUs = DEFAULT_PERIOD_US;

    // Sensor-thread state
    private final double[] batch = new double[MAX_BATCH_SAMPLES * 4];
    private int batchCount = 0;
    private double batchOffsetMs = 0;
    private double flushSpanMs = DEFAULT_BATCH_MS - 10;
    private int clockBase = CLOCK_UNKNOWN;

    private void ensureSensor() {
        if (sensorManager != null) return;
        sensorManager = (SensorManager) getContext().getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager != null) {
            sensor = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD_UNCALIBRATED);
        }
    }

    @PluginMethod
    public void getCapabilities(PluginCall call) {
        ensureSensor();
        JSObject ret = new JSObject();
        ret.put("version", VERSION);
        ret.put("available", sensor != null);
        ret.put("uncalibrated", sensor != null);
        if (sensor != null) {
            ret.put("minDelayUs", sensor.getMinDelay());
            ret.put("maxRange", sensor.getMaximumRange());
            ret.put("resolution", sensor.getResolution());
            ret.put("name", sensor.getName());
            ret.put("vendor", sensor.getVendor());
        }
        call.resolve(ret);
    }

    @PluginMethod
    public void start(PluginCall call) {
        ensureSensor();
        if (sensorManager == null || sensor == null) {
            call.reject("Uncalibrated magnetometer not available", "UNAVAILABLE");
            return;
        }

        int minDelay = sensor.getMinDelay();
        if (minDelay <= 0) minDelay = DEFAULT_PERIOD_US;
        int requested = call.getInt("periodUs", DEFAULT_PERIOD_US);
        periodUs = Math.max(requested, minDelay);
        int batchMs = Math.min(Math.max(call.getInt("batchMs", DEFAULT_BATCH_MS), 20), 500);
        flushSpanMs = Math.max(batchMs - 10, 10);

        // Restart cleanly if already running
        sensorManager.unregisterListener(this);
        resetPending = true;

        Looper looper = Looper.myLooper();
        handler = new Handler(looper != null ? looper : Looper.getMainLooper());
        // maxReportLatencyUs = 0: no hardware batching, every event as it happens.
        boolean ok = sensorManager.registerListener(this, sensor, periodUs, 0, handler);
        if (!ok) {
            active = false;
            call.reject("Could not register the magnetometer listener", "REGISTER_FAILED");
            return;
        }
        active = true;
        call.resolve();
    }

    @PluginMethod
    public void stop(PluginCall call) {
        active = false;
        if (sensorManager != null) {
            sensorManager.unregisterListener(this);
        }
        resetPending = true;
        call.resolve();
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (!active) return;

        if (resetPending) {
            resetPending = false;
            batchCount = 0;
            clockBase = CLOCK_UNKNOWN;
        }

        long ts = event.timestamp;
        if (clockBase == CLOCK_UNKNOWN) {
            clockBase = detectClockBase(ts);
        }

        double tMs;
        if (clockBase == CLOCK_ARRIVAL) {
            tMs = System.currentTimeMillis();
        } else {
            // One wall-clock offset per batch: per-sample offsets would add the jitter of
            // reading two clocks to every sample.
            if (batchCount == 0) {
                long nowWall = System.currentTimeMillis();
                long nowBase = clockBase == CLOCK_BOOT
                    ? SystemClock.elapsedRealtimeNanos()
                    : SystemClock.uptimeMillis() * 1_000_000L;
                batchOffsetMs = nowWall - nowBase / 1e6;
            }
            tMs = ts / 1e6 + batchOffsetMs;
        }

        int i = batchCount * 4;
        batch[i] = tMs;
        // values[0..2] = uncalibrated field; values[3..5] would be the estimated bias.
        batch[i + 1] = event.values[0];
        batch[i + 2] = event.values[1];
        batch[i + 3] = event.values[2];
        batchCount++;

        if (tMs - batch[0] >= flushSpanMs || batchCount >= MAX_BATCH_SAMPLES) {
            flush();
        }
    }

    private int detectClockBase(long ts) {
        // Since API 24 (minSdk) sensor timestamps must be on elapsedRealtimeNanos; some
        // older HALs used uptime. Anything else falls back to arrival time.
        long bootLag = SystemClock.elapsedRealtimeNanos() - ts;
        if (bootLag >= 0 && bootLag < CLOCK_MATCH_NS) return CLOCK_BOOT;
        long uptimeLag = SystemClock.uptimeMillis() * 1_000_000L - ts;
        if (uptimeLag >= 0 && uptimeLag < CLOCK_MATCH_NS) return CLOCK_UPTIME;
        return CLOCK_ARRIVAL;
    }

    private void flush() {
        if (batchCount == 0) return;
        JSArray d = new JSArray();
        try {
            for (int i = 0; i < batchCount * 4; i++) {
                d.put(batch[i]);
            }
        } catch (JSONException e) {
            // Only thrown for NaN/Infinity, which a sensor should never produce.
            batchCount = 0;
            return;
        }
        batchCount = 0;

        JSObject payload = new JSObject();
        payload.put("v", VERSION);
        payload.put("clk", clockBase == CLOCK_BOOT ? "boot" : clockBase == CLOCK_UPTIME ? "uptime" : "arrival");
        payload.put("d", d);
        notifyListeners("samples", payload);
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Uncalibrated values carry no calibration accuracy worth acting on.
    }

    // The sensor must not run while the app is in the background (battery). The TS layer
    // also stops on visibility change; this covers the gap before JS gets to run.
    @Override
    protected void handleOnPause() {
        super.handleOnPause();
        if (sensorManager != null && active) {
            sensorManager.unregisterListener(this);
            resetPending = true;
        }
    }

    @Override
    protected void handleOnResume() {
        super.handleOnResume();
        if (sensorManager != null && sensor != null && active && handler != null) {
            resetPending = true;
            sensorManager.registerListener(this, sensor, periodUs, 0, handler);
        }
    }

    // MainActivity.recreate() (after a renderer crash) builds a new plugin instance; the
    // old one must not keep a sensor listener alive.
    @Override
    protected void handleOnDestroy() {
        active = false;
        if (sensorManager != null) {
            sensorManager.unregisterListener(this);
        }
        super.handleOnDestroy();
    }
}
